emulate-browsers/base/lib/getLocalOperatingSystemMeta.ts

```diff
@@ -0,0 +1,112 @@
+import Os from 'os';
+import { convertVersionsToTree, getClosestNumberMatch } from "./VersionUtils";
+
+export default function getLocalOperatingSystemMeta() {
+  const platform = Os.platform();
+  const release = Os.release();
+  const name = extractOsName(platform);
+  const version = extractOsVersion(platform, release);
+  return { name, version };
+}
+
+function extractOsName(platform: string) {
+  return platformToOsName[platform.toLowerCase()];
+}
+
+function extractOsVersion(platform: string, release: string) {
+  let versionStr = '';
+
+  if (platform === 'darwin') {
+    versionStr = extractMacOsVersion(release);
+  } else if (platform === 'win32') {
+    versionStr = extractWindowsVersion(release);
+  } // else if linux then no version
+
+  return versionStr.split('.').slice(0, 2).join('-');
+}
+
+function extractWindowsVersion(release: string) {
+  // there is no guarantee we have an exact match, so let's get the closest
+  const [major, minor] = release.split('.');
+  const majors = Object.keys(windowsVersionTree).map(x => Number(x));
+  const majorMatch = getClosestNumberMatch(Number(major), majors);
+  let versionMatch = `${majorMatch}`;
+
+  const minors = Object.keys(windowsVersionTree[majorMatch]).map(x => Number(x));
+  const minorMatch = getClosestNumberMatch(Number(minor), minors);
+  versionMatch += `.${minorMatch}`;
+
+  return windowsToWindowsVersionMap[versionMatch];
+}
+
+function extractMacOsVersion(release: string) {
+  // there is no guarantee we have an exact match, so let's get the closest
+  const [major, minor, build] = release.split('.');
+  const majors = Object.keys(darwinVersionTree).map(x => Number(x));
+  const majorMatch = getClosestNumberMatch(Number(major), majors);
+  let versionMatch = `${majorMatch}`;
+
+  const minors = Object.keys(darwinVersionTree[majorMatch]).map(x => Number(x));
+  const minorMatch = getClosestNumberMatch(Number(minor), minors);
+  versionMatch += `.${minorMatch}`;
+
+  if (build) {
+    const builds = darwinVersionTree[majorMatch][minorMatch];
+    const buildMatch = getClosestNumberMatch(Number(build), builds);
+    versionMatch += `.${buildMatch}`;
+  }
+
+  return darwinToMacOsVersionMap[versionMatch];
+}
+
+const platformToOsName = {
+  darwin: 'mac-os',
+  win32: 'windows',
+  linux: 'linux',
+  aix: 'linux',
+  freebsd: 'linux',
+  openbsd: 'linux',
+  sunos: 'linux',
+}
+
+const darwinToMacOsVersionMap = {
```

Comment:
Any thoughts on how to make sure this stays up to date? Is anyone tracking this in a high usage, no deps lib?? :)

Reply:
I just pushed a commit that includes the wiki URL where I grabbed the data. I didn't see a dep lib, and I'm not certain of the benefits of trusting a 3rd party for something that changes maybe once a year.

All the version matching (when getting the local operating system and when getting the dom diffs) use `getClosestNumberMatch` in VersionUtils.ts. This method gets the closest version match without requiring an exact match. This means, developers can upgrade to a new OS version without our code needing to change.

Reply:
Ok. Is there some kind of warning or error if we can't match? If this is a year apart for changes, seems very easy to forget to update it.

Reply:
No. I'm not sure where to put the warning. I'm against clogging up Stdout with these types of messages.

Reply:
If there's no warning, and no message to console, this seems very likely, imo, to be lost. Could it be part of the shapeshifter somehow?

Reply:
Maybe. Not quite sure how yet.

Reply:
I'm not sure pulling it into Shapeshifter solves anything...

The data is still manually gathered... so I'm not sure how we would be "automatically" notified when the data is out of data. We might be able to solve this when we add "analytics" to SA.